Add profile_image information to Accounts API

dan-f opened the pull request:
Adds read-only profile_image metadata to responses from the Accounts API.

The first commit fixes a bug in the generation of migrations on the student module. Once the corresponding PR (https://github.com/edx/edx-platform/pull/7283) is merged, the commit should be removed from this branch.

TNL-1545

@jimabramson

jimabramson commented:
@dan-f code looks nice, though i still managed to find a bunch of things to complain about.  (also looks like there are test issues)

cahrens commented:
@dan-f I would like to review this PR also. Please let me know when you've updated it for Jim's feedback.

dan-f commented:
Sure thing, @cahrens

dan-f commented:
@jimabramson - I addressed your feedback. @cahrens I'm going to implement the test skipping logic you mentioned but feel free to take a look now if you'd like.

cahrens commented on common/djangoapps/student/migrations/0047_add_has_profile_image_boolean.py:
```diff
@@ -0,0 +1,194 @@
+# -*- coding: utf-8 -*-
+from south.utils import datetime_utils as datetime
```
We will need to figure how to handle the fact that we have "multiple 47" migration files.

dan-f replied:
I asked @cpennington about this.

dan-f replied:
It looks like whichever one of us lands second will have to re-generate the migrations.

cahrens replied:
So should you regenerate this one since learner-profiles already contains a "47" migration?

cahrens commented on lms/envs/common.py:
```diff
@@ -304,7 +304,7 @@
 
     # Flag to enable new user account APIs.
-    'ENABLE_USER_REST_API': False,
+    'ENABLE_USER_REST_API': True,
```
Since this is a feature branch, I am OK with turning this to True. Ultimately we will want to completely delete ENABLE_USER_REST_API.

cahrens commented on lms/envs/common.py:
```diff
@@ -2088,3 +2088,19 @@
 }
+
+# PROFILE IMAGE CONFIG
+# TODO: add these settings to aws.py as well
```
Please have someone from devops review this configuration setting so we know where it will need to live. If we need to add things to aws.py, let's do it now.

dan-f replied:
Thanks. @e0d could you please let me know if I need to move these settings?

cahrens commented on openedx/core/djangoapps/user_api/accounts/helpers.py:
```diff
@@ -0,0 +1,56 @@
+from django.conf import settings
+from django.core.files.storage import FileSystemStorage, get_storage_class
+
+PROFILE_IMAGE_SIZES = {
```
Are we supporting all these sizes now?

jimabramson replied:
@cahrens we'll generate all the sizes now, since doing less initially would leave people with an incomplete set later on.  I suppose it's not strictly necessary to give all their URLs from the API now, but not sure if there's any reason not to.

jimabramson commented on lms/envs/common.py:
```diff
@@ -2088,3 +2088,19 @@
+# i.e. 'http://www.example-image-server.com/'
+PROFILE_IMAGE_DOMAIN = '/'
+PROFILE_IMAGE_URL_PATH = 'media/profile_images/'
+PROFILE_IMAGE_DEFAULT_FILENAME = 'default_profile_image'  # TODO: determine final name
```
I think we need a comment here advising the operator that they will have to manually put the files in place when they set up the openedx instance, assuming you agree that's a reasonable setup step.  (I think trying to get the code to do something clever and automatically install files might be more complexity than it's worth.)

dan-f replied:
I agree with not doing fancy file installation stuff, especialy since that seems to cross into theming territory. Since I also don't think we should provide a default image to the open edx community, I'll advocate for just having a comment here.

jimabramson replied:
@e0d @maxrothman FYI for configurations of ephemeral environments (devstack/fullstack, sandboxes) would be best to copy the placeholder images into place as part of provisioning.  e.g. in the filesystem-backed scenario we'd move some assets from a place in the code repo over to `/edx/var/edxapp/{PROFILE_IMAGE_URL_PATH}`.  I assume we need to file a ticket for ansible changes to do this.  Please advise if we should be specifying settings any differently than they are here in order to make that happen.

cahrens commented on openedx/core/djangoapps/user_api/accounts/tests/test_helpers.py:
```diff
@@ -0,0 +1,54 @@
+
+@ddt
+@patch.dict(
+    'openedx.core.djangoapps.user_api.accounts.helpers.PROFILE_IMAGE_SIZES', {'full': 50}, clear=True
```
Add a second size.

cahrens commented on openedx/core/djangoapps/user_api/accounts/tests/test_helpers.py:
```diff
@@ -0,0 +1,54 @@
+        Verify we get the URL to the default image if the user does not
+        have a profile image.
+        """
+        self.user.has_profile_image = False
```
If we create a new user, we expect has_profile_mage to be False. Let's test that.

Then test setting it to True, and then back to False. I feel like you should be saving though -- user.profile.save().

jimabramson replied:
in this test case we set `self.user.has_profile_image` but in the next one we use `self.user.profile.has_profile_image`.  I think this one is incorrect?

jimabramson replied:
...but since it default to False the test passes anyway

cahrens commented on openedx/core/djangoapps/user_api/accounts/tests/test_views.py:
```diff
@@ -74,7 +75,10 @@ def create_mock_profile(self, user):
 @unittest.skipUnless(settings.ROOT_URLCONF == 'lms.urls', 'Account APIs are only supported in LMS')
-class TestAccountAPI(UserAPITestCase):
+@patch.dict(
+    'openedx.core.djangoapps.user_api.accounts.helpers.PROFILE_IMAGE_SIZES', {'full': 50}, clear=True
```
Have at least 2 sizes.

cahrens commented on openedx/core/djangoapps/user_api/accounts/tests/test_views.py:
```diff
@@ -83,6 +87,21 @@ def setUp(self):
+    def _verify_profile_image_data(self, data):
+        """
+        Verify the profile image data in a GET response for self.user is
+        what we expect.  Note that we mocked out the URL generation
```
By "what we expect", this is always checking that the user does not have a custom profile_image. Either rename the method so it is clear (and update doc), or have the method handle the other case as well. I assume someplace we are going to test what is returned when the user does have a custom profile image set.

jimabramson commented on openedx/core/djangoapps/user_api/accounts/helpers.py:
```diff
@@ -0,0 +1,56 @@
+    Raises:
+        ValueError: The caller asked for an unsupported image size.
+    """
+    if size not in PROFILE_IMAGE_SIZES.values():
```
instead of doing `.values()` on the dict (which creates a new list) each time this is called, instead, create a list from the values as a module variable and check that.

```
$ python -m timeit -s 'v = {"full": 500, "large": 150, "medium": 50, "small": 30}' '1 in v.values()'
1000000 loops, best of 3: 0.256 usec per loop
$ python -m timeit -s 'v = [500, 150, 50, 30]' '1 in v'
10000000 loops, best of 3: 0.115 usec per loop
```

dan-f replied:
This is a little tricky... I could separate those values out into a list or tuple, but then I'd have something like...

```python
PROFILE_IMAGE_SIZES = (500, 120, 50, 30)
PROFILE_IMAGE_SIZES_MAP = {
    'full': PROFILE_IMAGE_SIZES[0],
    'large': PROFILE_IMAGE_SIZES[1],
    'medium': PROFILE_IMAGE_SIZES[2],
    'small': PROFILE_IMAGE_SIZES[3]
}
```

... since we still need a way of mapping keys for the API response to URLs containing those pixel sizes. A side-effect of that approach is that overriding these settings becomes more cumbersome when testing.

Or were you suggesting something like this, where the list is still computed, but at module load time?:
```python
PROFILE_IMAGE_SIZES_MAP = {
    'full': PROFILE_IMAGE_SIZES[0],
    'large': PROFILE_IMAGE_SIZES[1],
    'medium': PROFILE_IMAGE_SIZES[2],
    'small': PROFILE_IMAGE_SIZES[3]
}
PROFILE_IMAGE_SIZES = PROFILE_IMAGE_SIZES_MAP.values()
```

jimabramson replied:
i was suggesting the latter.   just to create the list once, instead of recreating repeatedly inside the function call.

jimabramson replied:
wait, sorry. should have read more closely.  what i was suggesting was not to change the `PROFILE_IMAGE_SIZES` construct but just to add something extra:
```
PROFILE_IMAGE_SIZES = {
    'full': 500,
    'large': 120,
    'medium': 50,
    'small': 30
}
_PROFILE_IMAGE_SIZE_VALUES = PROFILE_IMAGE_SIZES.values()
```
and check `_PROFILE_IMAGE_SIZE_VALUES` in the subsequent function calls.

cahrens commented on openedx/core/djangoapps/user_api/accounts/tests/test_views.py:
```diff
@@ -103,14 +122,14 @@ def _verify_private_account_response(self, response):
-        self.assertEqual(12, len(data))
+        self.assertEqual(13, len(data))
         self.assertEqual(self.user.username, data["username"])
         self.assertEqual(self.user.first_name + " " + self.user.last_name, data["name"])
```
Add check of profile image as well. Perhaps update "mock profile" method so it ensures that the user DOES have a custom profile image?

cahrens commented on openedx/core/djangoapps/user_api/accounts/tests/test_views.py:
```diff
@@ -224,7 +243,7 @@ def test_get_account_default(self):
         response = self.send_get(self.client)
         data = response.data
-        self.assertEqual(12, len(data))
+        self.assertEqual(13, len(data))
```
Again, whenever you bump up the number of fields, you should also be testing that the profile image information is correct.

cahrens commented on openedx/core/djangoapps/user_api/accounts/tests/utils.py:
```diff
@@ -0,0 +1,20 @@
+from django.test.utils import override_settings
+
+
+@override_settings(
```
An alternative to this mixin is to just put the configuration in lms/envs/test.py. I think that would be better because that is where we usually put test-specific configuration, this mixin does nothing else.

dan-f replied:
That's much simpler. Thanks!

cahrens commented on openedx/core/djangoapps/user_api/accounts/views.py:
```diff
@@ -111,6 +111,10 @@ def get(self, request, username):
         try:
             account_settings = get_account_settings(request.user, username, view=request.QUERY_PARAMS.get('view'))
+            # Account for possibly relative URLs.
+            for key, value in account_settings['profile_image'].items():
```
This behavior of converting the URL to an absolute uri should have unit test.

cahrens commented on openedx/core/djangoapps/user_api/accounts/views.py:
```diff
@@ -111,6 +111,10 @@ def get(self, request, username):
             account_settings = get_account_settings(request.user, username, view=request.QUERY_PARAMS.get('view'))
+            # Account for possibly relative URLs.
+            for key, value in account_settings['profile_image'].items():
+                if key.startswith('image_url'):
```
Make "image_url" a constant.

cahrens commented:
I'm done reviewing @dan-f-- generally good, just a few things to clean up.

Note that we don't currently show test coverage for things in openedx. @andy-armstrong knows the current status of that, but in the meantime, you have to be extra diligent about making sure the new logic is all covered with tests.

andy-armstrong commented:
@dan-f @cahrens - I merged a PR to master that adds openedx to diff cover, so if you rebase you should pick it up. It is definitely working as you can see it on my PR:

https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/1705/Diff_Coverage_Report/

dan-f commented:
@cahrens @jimabramson - I've responded to your feedback in my latest fixup commit.

cahrens commented on openedx/core/djangoapps/user_api/tests/test_views.py:
```diff
@@ -1517,8 +1517,8 @@ def _assert_reg_field(self, extra_fields_setting, expected_field):
 
-
 @ddt.ddt
+@skipUnless(settings.ROOT_URLCONF == 'lms.urls', 'Account APIs are only supported in LMS')
```
I'm still confused about why this started failing (the email opt in has its own endpoint, and actually is available in CMS). Can you check again to see why this is necessary?

cahrens commented:
:+1: after you have re-enabled the email opt in test. I also want to make sure that someone from dev ops reviews the various settings, but that could be done later if necessary (since this is going on a feature branch).

dan-f commented:
@jimabramson can I get a thumbsup ?

dan-f commented:
@cahrens regarding the settings - these may shift around as @jimabramson builds on top of my work, so I'll vote for waiting for things to settle before we bother devops.

jimabramson commented:
------------> :+1:
